### PR TITLE
arch: aarch64: do EL2 init in EL3 if necessary

### PR DESCRIPTION
**Pull request by jharris-intel**
If EL2 is implemented but we're skipping EL2, we should still
do EL2 init. Otherwise we end up with a bunch of things still
at their (unknown) reset values.

This in particular causes problems when different
cores have different virtual timer offsets.

Signed-off-by: James Harris <james.harris@intel.com>

### CODE REVIEW

**Comment by jharris-intel** [comment]
@MrVan - FYI.

File: arch/arm/core/aarch64/reset.c [hunk 1]
```diff
@@ -36,6 +38,28 @@ void z_arm64_el_highest_init(void)
...
+	EL3_TO_EL1_SKIP_EL2
+};
+
+static inline enum el3_next_el arm64_el3_get_next_el(void)
+{
+	/*
+	 * Is considered an illegal return "[..] a return to EL2 when EL3 is
+	 * implemented and the value of the SCR_EL3.NS bit is 0 if
+	 * ARMv8.4-SecEL2 is not implemented" (D1.11.2 from ARM DDI 0487E.a)
+	 */
```
Comment (carlocaione) [inline arch/arm/core/aarch64/reset.c:53]: nit: can you move this comment to the proper `EL3_TO_EL1_SKIP_EL2` case?

File: arch/arm/core/aarch64/reset.c [hunk 2]
```diff
@@ -64,6 +88,14 @@ void z_arm64_el3_init(void)
 	z_arm64_el3_plat_init();
 
 	isb();
+
+	if (arm64_el3_get_next_el() == EL3_TO_EL1_SKIP_EL2) {
+		/* handle EL2 init in EL3,
+		 * as it still needs to be done
+		 * but we are going to be skipping EL2.
+		 */
```
Comment (carlocaione) [inline arch/arm/core/aarch64/reset.c:96]: nit: I know that this is not mandate in the Zephyr coding style but can you change the multi-line comment to the linux kernel coding style?
```
/*
 * comment
 * comment
 */
```
and also fix the length of the lines.

**Comment by carlocaione** [comment]
With the nits fixed LGTM.

**Review by carlocaione — APPROVED** [review]

File: arch/arm/core/aarch64/reset.c (hunk 1 above)
Comment (jharris-intel) [reply arch/arm/core/aarch64/reset.c:53]: Sure? I figured that the logic made more sense at the start of the thing that was checking "when EL3 is implemented and the value of the SCR_EL3.NS bit is 0 if ARMv8.4-SecEL2 is not implemented".

Comment (jharris-intel) [reply arch/arm/core/aarch64/reset.c]: Done.

File: arch/arm/core/aarch64/reset.c (hunk 2 above)
Comment (jharris-intel) [reply arch/arm/core/aarch64/reset.c]: Done.

**Comment by jharris-intel** [comment]
> GitHub Actions has encountered an internal error when running your job.

um.

File: arch/arm/core/aarch64/reset.c [hunk 3]
```diff
@@ -36,6 +38,28 @@ void z_arm64_el_highest_init(void)
 	isb();
 }
 
+enum el3_next_el {
+	EL3_TO_EL2,
+	EL3_TO_EL1_NO_EL2,
+	EL3_TO_EL1_SKIP_EL2
+};
+
+static inline enum el3_next_el arm64_el3_get_next_el(void)
```
Comment (carlocaione) [inline arch/arm/core/aarch64/reset.c:47]: @jharris-intel since you have to push again anyway to unstuck CI, can you rename this to `el3_get_next_el()`?

**Comment by carlocaione** [comment]
> Is trunk broken or something?

yes

**Review by MrVan — APPROVED** [review]

**Pull request merged**